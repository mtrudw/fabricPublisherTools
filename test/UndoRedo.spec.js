import {fabric} from 'fabric';
import {fabricAddUndoRedo} from '../src/index.js';

var canvas;

beforeAll(() => {
    fabricAddUndoRedo();
    document.body.innerHTML =
	'<canvas id="c"></canvas>'

});

beforeEach(() => {
    canvas = new fabric.Canvas('c');
});

test('inits History', () => {
    expect(canvas.id).toBeDefined();
    expect(canvas.historyStack).toBeDefined();
    expect(canvas.redoStack).toBeDefined();
    expect(canvas.recordHistory).toBeTruthy();
    expect(canvas.redoing).toBeFalsy();
    
});

test('pushes to History Stack', () => {
    var event ='test';
    canvas.recordHistory = true;
    canvas._pushToHistory(event);
    expect(canvas.historyStack).toContain('test');

    canvas.recordHistory = false;
    event = 'test2';
    canvas._pushToHistory(event);
    expect(canvas.historyStack).toContain('test');
    expect(canvas.historyStack).not.toContain('test2');
});

test('test add event undo and redo',() => {
    canvas.historyStack=[];
    canvas.redoStack=[];
    var rect = new fabric.Rect()

    canvas.add(rect);
    var jsonInit = canvas.toJSON();
    expect(jsonInit).toMatchSnapshot();
    
    canvas.undo();
    var jsonUndo = canvas.toJSON();
    expect(jsonUndo).toMatchSnapshot();

    canvas.redo();
    var jsonRedo = canvas.toJSON();
    expect(jsonRedo).toEqual(jsonInit);
    
});

test('test remove event undo and redo',() => {
    canvas.historyStack=[];
    canvas.redoStack=[];
    var rect = new fabric.Rect()

    canvas.add(rect);
    canvas.remove(rect);
    var jsonInit = canvas.toJSON();
    expect(jsonInit).toMatchSnapshot();
    
    canvas.undo();
    var jsonUndo = canvas.toJSON();
    expect(jsonUndo).toMatchSnapshot();

    canvas.redo();
    var jsonRedo = canvas.toJSON();
    expect(jsonRedo).toEqual(jsonInit);
    
});

test('test object modified undo and redo', () => {
    var rect = new fabric.Rect({width: 10,height: 10,left:  0,	top: 0, });
    var event ={"e":{"isTrusted":true},"target":{"type":"rect","version":"3.6.3","originX":"left","originY":"top","left":205.33,"top":22.67,"width":10,"height":10,"fill":"rgb(0,0,0)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeDashOffset":0,"strokeLineJoin":"miter","strokeMiterLimit":4,"scaleX":1,"scaleY":1,"angle":0,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","paintFirst":"fill","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"rx":0,"ry":0,"id":0},"transform":{"target":{"type":"rect","version":"3.6.3","originX":"left","originY":"top","left":205.33,"top":22.67,"width":10,"height":10,"fill":"rgb(0,0,0)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeDashOffset":0,"strokeLineJoin":"miter","strokeMiterLimit":4,"scaleX":1,"scaleY":1,"angle":0,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","paintFirst":"fill","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"rx":0,"ry":0,"id":0},"action":"drag","corner":"bl","scaleX":1,"scaleY":1,"skewX":0,"skewY":0,"offsetX":2.4166695351646537,"offsetY":7.933341082011751,"originX":"right","originY":"top","ex":2.4166695351646537,"ey":7.933341082011751,"lastX":2.4166695351646537,"lastY":7.933341082011751,"theta":0,"width":10,"mouseXSign":1,"mouseYSign":1,"shiftKey":false,"altKey":false,"original":{"scaleX":1,"scaleY":1,"skewX":0,"skewY":0,"angle":0,"left":0,"flipX":false,"flipY":false,"top":0,"originX":"right","originY":"top"},"reset":false,"actionPerformed":true}}

    canvas.historyStack=[];
    canvas.redoStack=[];

    canvas.add(rect);
    var jsonInit = canvas.toJSON();
    expect(jsonInit).toMatchSnapshot();
    
    canvas.trigger('object:modified',event);
    var jsonEvent = canvas.toJSON();
    expect(jsonEvent).toMatchSnapshot();

    canvas.undo();
    var jsonUndo = canvas.toJSON();
    expect(jsonUndo).toMatchSnapshot();

    canvas.redo();
    var jsonRedo = canvas.toJSON();
    expect(jsonRedo).toMatchSnapshot();

});


test('test object set method', () => {
    canvas.historyStack=[];
    canvas.redoStack=[];
    var rect = new fabric.Rect()

    canvas.add(rect);
    var jsonInit = canvas.toJSON();
    expect(jsonInit).toMatchSnapshot();

    rect.set('top',100);
    var jsonUnchanged = canvas.toJSON();
    expect(jsonUnchanged.historyStack).toEqual(jsonInit.historyStack);

    rect.set('fill','#FFF');
    var jsonChangeFill = canvas.toJSON();
    expect(jsonChangeFill).toMatchSnapshot;
})
