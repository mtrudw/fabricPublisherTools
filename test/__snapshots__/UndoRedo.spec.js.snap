// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test add event undo and redo 1`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 0,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 0,
    },
  ],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test add event undo and redo 2`] = `
Object {
  "historyStack": Array [],
  "id": 1,
  "objects": Array [],
  "redoStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "version": "3.6.3",
}
`;

exports[`test object modified undo and redo 1`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 10,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 10,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 10,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 10,
    },
  ],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test object modified undo and redo 2`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 10,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 10,
      },
      "before": Array [],
      "id": 0,
    },
    Object {
      "action": "drag",
      "after": Array [
        Object {
          "left": 205.33,
        },
        Object {
          "top": 22.67,
        },
      ],
      "before": Array [
        Object {
          "left": 0,
        },
        Object {
          "top": 0,
        },
      ],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 10,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 10,
    },
  ],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test object modified undo and redo 3`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 10,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 10,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 10,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 10,
    },
  ],
  "redoStack": Array [
    Object {
      "action": "drag",
      "after": Array [
        Object {
          "left": 205.33,
        },
        Object {
          "top": 22.67,
        },
      ],
      "before": Array [
        Object {
          "left": 0,
        },
        Object {
          "top": 0,
        },
      ],
      "id": 0,
    },
  ],
  "version": "3.6.3",
}
`;

exports[`test object modified undo and redo 4`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 10,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 10,
      },
      "before": Array [],
      "id": 0,
    },
    Object {
      "action": "drag",
      "after": Array [
        Object {
          "left": 205.33,
        },
        Object {
          "top": 22.67,
        },
      ],
      "before": Array [
        Object {
          "left": 0,
        },
        Object {
          "top": 0,
        },
      ],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 10,
      "id": 0,
      "left": 205.33,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 22.67,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 10,
    },
  ],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test object set method 1`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 0,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 0,
    },
  ],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test remove event undo and redo 1`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "before": Array [],
      "id": 0,
    },
    Object {
      "action": "removed",
      "after": Array [],
      "before": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [],
  "redoStack": Array [],
  "version": "3.6.3",
}
`;

exports[`test remove event undo and redo 2`] = `
Object {
  "historyStack": Array [
    Object {
      "action": "added",
      "after": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "before": Array [],
      "id": 0,
    },
  ],
  "id": 1,
  "objects": Array [
    Object {
      "angle": 0,
      "backgroundColor": "",
      "clipTo": null,
      "fill": "rgb(0,0,0)",
      "fillRule": "nonzero",
      "flipX": false,
      "flipY": false,
      "globalCompositeOperation": "source-over",
      "height": 0,
      "id": 0,
      "left": 0,
      "opacity": 1,
      "originX": "left",
      "originY": "top",
      "paintFirst": "fill",
      "rx": 0,
      "ry": 0,
      "scaleX": 1,
      "scaleY": 1,
      "shadow": null,
      "skewX": 0,
      "skewY": 0,
      "stroke": null,
      "strokeDashArray": null,
      "strokeDashOffset": 0,
      "strokeLineCap": "butt",
      "strokeLineJoin": "miter",
      "strokeMiterLimit": 4,
      "strokeWidth": 1,
      "top": 0,
      "transformMatrix": null,
      "type": "rect",
      "version": "3.6.3",
      "visible": true,
      "width": 0,
    },
  ],
  "redoStack": Array [
    Object {
      "action": "removed",
      "after": Array [],
      "before": Object {
        "angle": 0,
        "backgroundColor": "",
        "clipTo": null,
        "fill": "rgb(0,0,0)",
        "fillRule": "nonzero",
        "flipX": false,
        "flipY": false,
        "globalCompositeOperation": "source-over",
        "height": 0,
        "id": 0,
        "left": 0,
        "opacity": 1,
        "originX": "left",
        "originY": "top",
        "paintFirst": "fill",
        "rx": 0,
        "ry": 0,
        "scaleX": 1,
        "scaleY": 1,
        "shadow": null,
        "skewX": 0,
        "skewY": 0,
        "stroke": null,
        "strokeDashArray": null,
        "strokeDashOffset": 0,
        "strokeLineCap": "butt",
        "strokeLineJoin": "miter",
        "strokeMiterLimit": 4,
        "strokeWidth": 1,
        "top": 0,
        "transformMatrix": null,
        "type": "rect",
        "version": "3.6.3",
        "visible": true,
        "width": 0,
      },
      "id": 0,
    },
  ],
  "version": "3.6.3",
}
`;
